### ISSUE TITLE
Read: xRefTable failed: pdfcpu: can't find last xref section

### ISSUE THREAD

**Issue by dhf0820** [pdfcpu/pdfcpu#99]
I have been using  Version: 0.1.18 for some time and ran into a problem of documents no merging and no error being reported.  I updated to the latest version, cant find the version number.

Now all documents are failing with the above error that have been working for the past year.
My command to merge was _, err := api.Process(api.MergeCommand(files, outFile, pdfcpu.NewDefaultConfiguration()))

Now those do not exist any more and I am using: err := cli.Process(cli.MergeCommand(files, outFile, pdfcpu.NewDefaultConfiguration())) and get the error on every document, EVEN ones you generate from tif combined images that have been working fro a year on the other release.

What has happened or am I doing something wrong?

Don French

**Comment by dhf0820** [comment]
I have a repeatable error on a document pdfcpu has just converted from tif. As in the past I need to transfer it to you securely.

**Comment by hhrutter** [comment]
You are doing nothing wrong. Like the ReadMe says pdfcpu is still Alpha.
With 0.2 the API was revamped and cleaned up.
Post the stacktrace of the repeatable error.

**Comment by hhrutter** [comment]
Any news on this?
If I don't hear back I am going to close this due to inactivity.

**Comment by dhf0820** [comment]
I did a work around with a ruby restful that converts until new release came out. I will check now. I do what it all in go.

Don French
702-503-9665

**Comment by hhrutter** [comment]
Closing due to inactivity.

**Comment by maxiride** [comment]
I'm having the same issue, I merged some pdfs and on a second run i got the error:
`Read: xRefTable failed: pdfcpu: can't find last xref section`

I'm doing the merging use the API in a codebase:

Several files are in an output folder (next to the executable) and I want the merged output to be placed next to them, the following snipped works on a first run, after that I have to delete the generated merged.pdf.

It was my understading that `pdfcupu.MergeCreateFile` [overwrites the destination file](https://godoc.org/github.com/pdfcpu/pdfcpu/pkg/api#MergeCreateFile)

```Go
	var ss []string
	for _, f := range files {
		ss = append(ss, "output/" + f.Name())
	}

	err := pdfcupu.MergeCreateFile(ss, "output/merged.pdf", nil)
	if err != nil {
		log.Println("Errore merging", err)
	}
```

**Comment by hhrutter** [comment]
Sorry I don't understand your issue.

This should help: https://github.com/pdfcpu/pdfcpu/blob/master/pkg/api/test/merge_test.go

If you do run into an error, please explain the details of your usecase 
eg. which command produces your error and by providing a stack trace .
Ideally I would need sample files for further analysis.

Thank you for using pdfcpu 💚

**Comment by maxiride** [comment]
Getting back a the issue, files sent via email.

The error occurred while applying stamps, either image or text.

The following functions are called synchronously in a for loop iterating over a lot of files, applying first an image, then three text strings in different places and confs.

The configurations:
```
conf: 'off: 180 240, points: 5, font: Helvetica, pos: bl, sc: 1 abs, rot: 0, fillc: #000000'
conf: 'off: 150 245, points: 8, font: Helvetica, pos: bl, sc: 1 abs, rot: 0, fillc: #000000'
conf: 'off: 150 240, points: 8, font: Helvetica, pos: bl, sc: 1 abs, rot: 0, fillc: #000000'
```

```
// pseudo code for brevity
for j := range jobs{
  applicaDatamatrix(j.file, j.dtmx)
  applicaTesto(j.file, j.t1, j.c1)
  applicaTesto(j.file, j.t1, j.c2)
  applicaTesto(j.file, j.t1, j.c3)
}


// applicaDatamatrix applica un dato file immagine datamatrix ad un dato file pdf, il nuovo file PDF viene salvato con lo stesso nome
// nel percorso di output con suffisso
// inFile, outFile e dtmxFile sono percorsi assoluti dei file di input, output e immagine del datamtrix
func (j *Job) applicaDatamatrix(inFile, dtmxFile string) error {
	err := pdfcpu.AddImageWatermarksFile(inFile, "", []string{"1"}, true, dtmxFile, j.confDtmx, imponi.PdfConf)
	if err != nil {
		return errors.Wrap(err, fmt.Sprintf("KO applicazione datamatrix %s", dtmxFile))
	}

	return nil
}

// applicaTesto applica del testo al PDF
func (j *Job) applicaTesto(inFile, text, conf string) error {
	err := pdfcpu.AddTextWatermarksFile(inFile, "", []string{"1"}, true, text, conf, imponi.PdfConf)
	if err != nil {
		return err
	}
	return nil
}
```

**Comment by hhrutter** [comment]
This issue is already closed and based on v0.1.18.
If you are using the old version please upgrade and check if the error persists.
If not, please open a new issue!

In any case I would apply a sequence of watermarks to an inFile like [in this test](https://github.com/pdfcpu/pdfcpu/blob/master/pkg/api/test/stampVersatile_test.go).

All 4 files you were sending are ok BTW.

**Comment by shubhagrwl** [comment]
this is how I solve this problem, hope this will help you

```
func addWatermarkImage(originalFileName, inputPath, watermark string) (fileBytes []byte, err error) {

	outFileName := originalFileName + "-watermark.pdf"
	outFileDir := filepath.Join(constants.FILEDIR, outFileName)
	conf := "fillc:#FF7A7A, font:Helvetica, points:24, position:c, offset:0 0, scale:0.5 rel, diagonal:1, opacity:1"

	var f1, f2 *os.File
	if f1, err = os.Open(inputPath); err != nil {
		return
	}

	if f2, err = os.Create(outFileDir); err != nil {
		return
	}

	wm, err := api.TextWatermark(watermark, conf, false, false, pdfcpu.POINTS)
	if err != nil {
		return
	}

	aerr := api.AddWatermarks(f1, f2, []string{"1-"}, wm, nil)

	// aerr := api.AddTextWatermarksFile(inputPath, outFileDir, []string{"1-"}, false, watermark, conf, nil)

	if aerr != nil {
		err = aerr
		return
	}
	if aerr = api.ValidateFile(outFileDir, nil); aerr != nil {
		err = aerr
		return
	}

	fileBytes, aerr = ioutil.ReadFile(outFileDir)
	if aerr != nil {
		err = aerr
		return
	}

	return fileBytes, nil
}

```

**Comment by Rupour** [comment]
Just in case someone else stumbles upon this, similar to maxiride I had this error because i was trying to merge pdfs and overwrite one of the files i was merging. Like as folllows:
"pdfcpu merge merged_file.pdf file1.pdf file2.pdf merged_file.pdf"

Similarly with the api MergeCreateFile:
pdfFiles := ["file1.pdf", "file2.pdf", "merged_file.pdf"]
pdfcupu.MergeCreateFile(pdfFiles, "merged_file.pdf", nil) 

For some reason this causes the XRefTable to fail. I tested this on v0.4.1 and v0.3.13. 

The only solution I found was to not overwrite the original file while including itself. 

Hopefully this helps!
